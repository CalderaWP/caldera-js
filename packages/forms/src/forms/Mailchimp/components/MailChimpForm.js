import React, {useState, useRef, useEffect} from "react";
import {PacmanLoader} from 'react-spinners';
import PropTypes from 'prop-types';
import {createSubscriber} from "../http/publicClient";
import {CalderaForm} from "../../../";

/**
 * Component for stand-alone mailchimp forms served via Caldera API
 *
 * @param form
 * @param onChange
 * @param onBlur
 * @param onSubmit
 * @return {*}
 * @constructor
 */
function MailChimpForm({form, onChange, onBlur, onSubmit, hideOnSubmit, onReady}) {
    const [isSubmitting, setIsSubmitting] = useState(false);
    const [completed, setIsCompleted] = useState(false);
    const [message, setMessage] = useState('');
    const readyRef = useRef(false);
    useEffect(() => {
        readyRef.current = true;
        Promise.resolve(onReady);
        return () => readyRef.current = false;
    }, [readyRef, onReady]);

    const Spinner = () => (
        <div><PacmanLoader/></div>
    );

    if (!form.hasOwnProperty('fields')) {
        return <Spinner/>
    }

    const {processors} = form;
    const processor = processors.find(p => 'mc-subscribe' === p.type);

    if (completed) {
        return <div className={'success'}>{message}</div>
    }

    if (isSubmitting) {
        return <Spinner/>;
    }

    return (
        <div>
            {message &&
            <div className={'error'}>{message}</div>
            }

            <CalderaForm
                form={form}
                onSubmit={(
                    //current values of all fields
                    values,
                    actions
                ) => {
                    setIsSubmitting(true);
                    onSubmit(values, processor).then(r => r.json()).then(r => {
                        setIsSubmitting(false);

                        if (hideOnSubmit) {
                            setMessage(r.message);
                            setIsCompleted(r);
                        }

                    })
                        .catch(e => {
                            setIsSubmitting(false);
                            if (e.hasOwnProperty('message')) {
                                setMessage(e.message);
                            } else {
                                setMessage('An Error happened.');

                            }
                        });
                }}
                onChange={(values) => {
                    onChange(values) //all field values
                }}
            />
        </div>

    )
}

/**
 * Default Submit handler
 *
 * @param values
 * @return {*}
 */
const onSubmit = (values, processor) => {
    return createSubscriber(values, processor);
};

MailChimpForm.propTypes = {
    listId: PropTypes.string,
    form: PropTypes.object,
    onSubmit: PropTypes.func,
    onChange: PropTypes.func,
    onBlur: PropTypes.func,
    onReady: PropTypes.instanceOf(Promise),
};

const noop = () => {
};
const promiseNoop = new Promise((resolve) => {
    resolve();
});

MailChimpForm.defaultProps = {
    onSubmit: noop,
    onChange: noop,
    hideOnSubmit: true,
    onReady: promiseNoop
};
export default MailChimpForm;
