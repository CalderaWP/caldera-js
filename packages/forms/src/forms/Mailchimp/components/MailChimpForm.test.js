import React from 'react';
import { render, fireEvent, getByTestId} from "react-testing-library";

import MailChimpForm from "./MailChimpForm";

import {mailChimpTestForm} from "./mailChimpTestForm.fixture";


describe( 'MailChimp mailChimpTestForm', () => {
	let onSubmit, onBlur,onChange;
	beforeEach( () => {
		onSubmit = jest.fn();
		onBlur = jest.fn();
		onChange = jest.fn();
	});

	it('Shows spinner if mailChimpTestForm does not have fields', () => {

		expect(render(
			<MailChimpForm
				form={{
					ID: 'mc-test-1'
				}}
				onBlur={onBlur}
				onChange={onChange}
				onSubmit={onSubmit}
			/>
		)).toMatchSnapshot();


	});

	it('Loads if it has proper mailChimpTestForm', () => {

		expect(render(
			<MailChimpForm
				form={mailChimpTestForm}
				onBlur={onBlur}
				onChange={onChange}
				onSubmit={onSubmit}
			/>
		)).toMatchSnapshot();


	});
});



