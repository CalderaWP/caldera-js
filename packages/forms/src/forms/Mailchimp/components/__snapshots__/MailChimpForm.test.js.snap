// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailChimp mailChimpTestForm Loads if it has proper mailChimpTestForm 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="css-16tm1g7-wrapper-wrapper"
        >
          <div
            class="css-1anlcob-pacmanStyle"
          />
          <div
            class="css-1oj2eaa-pacmanStyle"
          />
          <div
            class="css-1y2ttxx-ballStyle"
          />
          <div
            class="css-im5qma-ballStyle"
          />
          <div
            class="css-10tjdos-ballStyle"
          />
          <div
            class="css-x3nw5t-ballStyle"
          />
        </div>
      </div>
    </div>
    <div>
      <div>
        
        <div
          class="caldera"
        >
          <form
            class="mc-45907f0c59 caldera-form"
          >
            <div
              class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
            >
              <div
                class="caldera-column width-1 sc-bdVaJa iHZvIS"
                id="mc-45907f0c59-1-1"
              >
                <div
                  class="caldera-field-group caldera-field-area-input form-group"
                >
                  <label
                    class="form-label"
                    for="mc-email"
                  >
                    Email
                     
                    <span>
                      *
                    </span>
                  </label>
                  <input
                    class="caldera-field caldera-field-text form-control"
                    id="mc-email"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="caldera-column width-1 sc-bdVaJa iHZvIS"
                id="mc-45907f0c59-1-2"
              >
                <div
                  class="caldera-field-group caldera-field-area-input form-group"
                >
                  <label
                    class="form-label"
                    for="FNAME"
                  >
                    First Name
                     
                  </label>
                  <input
                    class="caldera-field caldera-field-text form-control"
                    id="FNAME"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
            >
              <div
                class="caldera-column width-1 sc-bdVaJa iHZvIS"
                id="mc-45907f0c59-2-1"
              >
                <div
                  class="caldera-field-group caldera-field-area-input form-group"
                >
                  <label
                    class="form-label"
                    for="LNAME"
                  >
                    Last Name
                     
                  </label>
                  <input
                    class="caldera-field caldera-field-text form-control"
                    id="LNAME"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="caldera-column width-1 sc-bdVaJa iHZvIS"
                id="mc-45907f0c59-2-2"
              >
                <div
                  class="caldera-field-group caldera-field-area-input form-group"
                >
                  <label
                    class="form-label"
                    for="MMERGE3"
                  >
                    Primary Website
                     
                  </label>
                  <input
                    class="caldera-field caldera-field-text form-control"
                    id="MMERGE3"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
            >
              <div
                class="caldera-column width-1 sc-bdVaJa iHZvIS"
                id="mc-45907f0c59-3-1"
              >
                <div
                  class="caldera-field-group caldera-field-area-checkboxes form-group"
                >
                  <label
                    class="form-label"
                    for="50c908e6aa"
                  >
                    What React Framework Do You Want To Learn?
                     
                  </label>
                  <fieldset>
                    <legend>
                      What React Framework Do You Want To Learn?
                    </legend>
                    <div
                      class="caldera-field caldera-field-checkbox form-check"
                    >
                      <input
                        class="form-check-input"
                        id="75091baaad"
                        type="checkbox"
                      />
                      <label
                        class="form-check-label"
                        for="75091baaad"
                        title=""
                        type="checkbox"
                      >
                        Gatsby
                      </label>
                    </div>
                    <div
                      class="caldera-field caldera-field-checkbox form-check"
                    >
                      <input
                        class="form-check-input"
                        id="0a4dc88500"
                        type="checkbox"
                      />
                      <label
                        class="form-check-label"
                        for="0a4dc88500"
                        title=""
                        type="checkbox"
                      >
                        Next.Js
                      </label>
                    </div>
                    <div
                      class="caldera-field caldera-field-checkbox form-check"
                    >
                      <input
                        class="form-check-input"
                        id="5a44f196ec"
                        type="checkbox"
                      />
                      <label
                        class="form-check-label"
                        for="5a44f196ec"
                        title=""
                        type="checkbox"
                      >
                        Create React App
                      </label>
                    </div>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
            >
              <div
                class="caldera-column width-1 sc-bdVaJa iHZvIS"
                id="mc-45907f0c59-4-2"
              >
                <div
                  class="caldera-field-group caldera-field-area-submit form-group"
                >
                  <label
                    class="form-label"
                    for="mc-submit"
                  >
                     
                  </label>
                  <input
                    class="caldera-field caldera-field-undefined form-control"
                    id="mc-submit"
                    value=""
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      
      <div
        class="caldera"
      >
        <form
          class="mc-45907f0c59 caldera-form"
        >
          <div
            class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
          >
            <div
              class="caldera-column width-1 sc-bdVaJa iHZvIS"
              id="mc-45907f0c59-1-1"
            >
              <div
                class="caldera-field-group caldera-field-area-input form-group"
              >
                <label
                  class="form-label"
                  for="mc-email"
                >
                  Email
                   
                  <span>
                    *
                  </span>
                </label>
                <input
                  class="caldera-field caldera-field-text form-control"
                  id="mc-email"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="caldera-column width-1 sc-bdVaJa iHZvIS"
              id="mc-45907f0c59-1-2"
            >
              <div
                class="caldera-field-group caldera-field-area-input form-group"
              >
                <label
                  class="form-label"
                  for="FNAME"
                >
                  First Name
                   
                </label>
                <input
                  class="caldera-field caldera-field-text form-control"
                  id="FNAME"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
          >
            <div
              class="caldera-column width-1 sc-bdVaJa iHZvIS"
              id="mc-45907f0c59-2-1"
            >
              <div
                class="caldera-field-group caldera-field-area-input form-group"
              >
                <label
                  class="form-label"
                  for="LNAME"
                >
                  Last Name
                   
                </label>
                <input
                  class="caldera-field caldera-field-text form-control"
                  id="LNAME"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="caldera-column width-1 sc-bdVaJa iHZvIS"
              id="mc-45907f0c59-2-2"
            >
              <div
                class="caldera-field-group caldera-field-area-input form-group"
              >
                <label
                  class="form-label"
                  for="MMERGE3"
                >
                  Primary Website
                   
                </label>
                <input
                  class="caldera-field caldera-field-text form-control"
                  id="MMERGE3"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
          >
            <div
              class="caldera-column width-1 sc-bdVaJa iHZvIS"
              id="mc-45907f0c59-3-1"
            >
              <div
                class="caldera-field-group caldera-field-area-checkboxes form-group"
              >
                <label
                  class="form-label"
                  for="50c908e6aa"
                >
                  What React Framework Do You Want To Learn?
                   
                </label>
                <fieldset>
                  <legend>
                    What React Framework Do You Want To Learn?
                  </legend>
                  <div
                    class="caldera-field caldera-field-checkbox form-check"
                  >
                    <input
                      class="form-check-input"
                      id="75091baaad"
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      for="75091baaad"
                      title=""
                      type="checkbox"
                    >
                      Gatsby
                    </label>
                  </div>
                  <div
                    class="caldera-field caldera-field-checkbox form-check"
                  >
                    <input
                      class="form-check-input"
                      id="0a4dc88500"
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      for="0a4dc88500"
                      title=""
                      type="checkbox"
                    >
                      Next.Js
                    </label>
                  </div>
                  <div
                    class="caldera-field caldera-field-checkbox form-check"
                  >
                    <input
                      class="form-check-input"
                      id="5a44f196ec"
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      for="5a44f196ec"
                      title=""
                      type="checkbox"
                    >
                      Create React App
                    </label>
                  </div>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
          >
            <div
              class="caldera-column width-1 sc-bdVaJa iHZvIS"
              id="mc-45907f0c59-4-2"
            >
              <div
                class="caldera-field-group caldera-field-area-submit form-group"
              >
                <label
                  class="form-label"
                  for="mc-submit"
                >
                   
                </label>
                <input
                  class="caldera-field caldera-field-undefined form-control"
                  id="mc-submit"
                  value=""
                />
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MailChimp mailChimpTestForm Shows spinner if mailChimpTestForm does not have fields 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="css-16tm1g7-wrapper-wrapper"
        >
          <div
            class="css-1anlcob-pacmanStyle"
          />
          <div
            class="css-1oj2eaa-pacmanStyle"
          />
          <div
            class="css-1y2ttxx-ballStyle"
          />
          <div
            class="css-im5qma-ballStyle"
          />
          <div
            class="css-10tjdos-ballStyle"
          />
          <div
            class="css-x3nw5t-ballStyle"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="css-16tm1g7-wrapper-wrapper"
      >
        <div
          class="css-1anlcob-pacmanStyle"
        />
        <div
          class="css-1oj2eaa-pacmanStyle"
        />
        <div
          class="css-1y2ttxx-ballStyle"
        />
        <div
          class="css-im5qma-ballStyle"
        />
        <div
          class="css-10tjdos-ballStyle"
        />
        <div
          class="css-x3nw5t-ballStyle"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
